c2rust-analyze/src/borrowck/mod.rs

```diff
@@ -15,25 +20,123 @@ mod dump;
 mod type_check;
 
 #[derive(Clone, Copy, PartialEq, Eq, Debug, Hash, Default)]
-pub struct Label {
+pub struct Label<'tcx> {
+    /// The [`Origin`] of this type
     pub origin: Option<Origin>,
+    /// The [Origins](`Origin`) associated with each lifetime
+    /// parameter of this type, if applicable
+    pub origin_params: Option<&'tcx [(OriginParam, Origin)]>,
     pub perm: PermissionSet,
 }
 
-pub type LTy<'tcx> = LabeledTy<'tcx, Label>;
-pub type LTyCtxt<'tcx> = LabeledTyCtxt<'tcx, Label>;
+pub type LTy<'tcx> = LabeledTy<'tcx, Label<'tcx>>;
+pub type LTyCtxt<'tcx> = LabeledTyCtxt<'tcx, Label<'tcx>>;
+
+/// Metadata describing lifetimes and lifetime parameters
+/// of a struct field.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FieldMetadata<'tcx> {
+    /// The lifetime parameters of a field, e.g. if a struct
+    /// `foo<'a, 'b>` is a field of `bar<'c, 'd>` as field: `foo<'c, 'd>`,
+    /// the lifetime params would be a set {'c, 'd}. For a reference such
+    /// as &'r foo<'c, 'd>, the lifetime params in the label would be
+    /// {'r}, and {'c, 'd} would be the label of the sole argument
+    /// of the labeled reference type
+    pub lifetime_params: LabeledTy<'tcx, &'tcx [OriginArg<'tcx>]>,
+}
+
+/// Metadata describing the lifetime parameters and fields
+/// of a struct.
+#[derive(Clone, PartialEq, Eq)]
+pub struct AdtMetadata<'tcx> {
+    /// The lifetime parameters of a structure, including
+    /// hypothetical lifetimes derived from pointer fields.
+    pub lifetime_params: IndexSet<OriginParam>,
+    pub field_info: IndexMap<DefId, FieldMetadata<'tcx>>,
+}
+
+impl Default for AdtMetadata<'_> {
+    fn default() -> Self {
+        Self {
+            lifetime_params: IndexSet::new(),
+            field_info: IndexMap::new(),
+        }
+    }
+}
+
+/// An origin parameter of a type to resolve in a MIR body
+/// that will get mapped to a concrete Origin to
+/// provide to polonius.
+#[derive(Hash, PartialEq, Eq, Clone, Copy)]
+pub enum OriginParam {
+    /// An existing [`EarlyBoundRegion`], i.e. `'a` in `struct A<'a>`
+    Actual(EarlyBoundRegion),
+    /// A hypothesized region derived from a pointer type
+    /// e.g. `'h0` derived from the pointer in `*mut foo`
+    Hypothetical(i64),
+}
+
+/// An origin parameter of a field type resolve in a MIR body
+/// that will get mapped to a concrete Origin to
+/// provide to polonius.
+#[derive(Hash, PartialEq, Eq, Clone, Copy)]
+pub enum OriginArg<'tcx> {
+    /// An existing [`Region`], i.e. `'a` in `&'a foo`.
+    /// Can be [RegionKind::ReEarlyBound](`rustc_type_ir::RegionKind`) or [RegionKind::ReStatic](`rustc_type_ir::RegionKind`)
+    Actual(Region<'tcx>),
+    /// A hypothesized region derived from a pointer type
+    /// e.g. `'h0` derived from the pointer in `*mut foo`
+    Hypothetical(i64),
+}
+
+impl<'tcx> std::fmt::Debug for OriginArg<'tcx> {
```

Comment:
```suggestion
impl<'tcx> Debug for OriginArg<'tcx> {
```

Can you `use std::fmt::Debug`?

Reply:
Elsewhere, too.